% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/eff_neighbor.R
\name{eff_neighbor}
\alias{eff_neighbor}
\title{Estimated self and neighbor QTL effects across a genome}
\usage{
eff_neighbor(
  genoprobs,
  pheno,
  gmap,
  contrasts = c(TRUE, TRUE, TRUE),
  smap,
  scale,
  addcovar = NULL,
  addQTL = NULL,
  grouping = rep(1, nrow(smap)),
  response = "quantitative",
  fig = TRUE
)
}
\arguments{
\item{genoprobs}{Conditional genotype probabilities as obtained from \code{qtl2::calc_genoprob()}.}

\item{pheno}{A vector of individual phenotypes.}

\item{gmap}{Genetic map including observed and pseudomarkers, as obtained from \code{qtl2::insert_pseudomarkers()}.}

\item{contrasts}{A vector composed of three TRUE/FALSE values. Depending on crossing design, it represents the presence/absence of specific genotypes as c(T/F, T/F, T/F) = AA, AB, BB.}

\item{smap}{A matrix showing a spatial map. The first and second column include spatial points along a x-axis and y-axis, respectively.}

\item{scale}{A numeric scalar indicating the maximum spatial distance between a focal individual and neighbors to define neighbor effects.}

\item{addcovar}{An optional matrix including additional non-genetic covariates. It contains no. of individuals x no. of covariates.}

\item{addQTL}{An optional vector containing marker names that are considered covariates. Namely, this option allows composite interval mapping (Jansen 1993).}

\item{grouping}{An optional integer vector assigning each individual to a group. This argument can be useful when \code{smap} contains different experimental replicates. Default setting means that all individuals are belong to a single group.}

\item{response}{An optional argument to select trait types. The \code{"quantitative"}, \code{"binary"}, or \code{"count"} calls the \code{"gaussian"}, \code{"binomial"}, or \code{"poisson"} family in \code{glm()}, respectively.}

\item{fig}{TRUE/FALSE to plot the effects or not.}
}
\value{
A matrix of estimated additive and dominance deviation for self and neighbor effects, with the chromosome numbers and positions. The row names correspond to marker names.
\itemize{
\item{\code{chr}} {Chromosome number}
\item{\code{pos}} {Marker position}
\item{\code{a1}} {Additive deviation for self effects}
\item{\code{d1}} {Dominance deviation for self effects}
\item{\code{a2}} {Additive deviation for neighbor effects}
\item{\code{d2}} {Dominance deviation for neighbor effects}
}
}
\description{
A function to estimate additive and dominance deviation for self and neighbor QTL effects by Haley-Knott regression
}
\examples{
#demo using F2
set.seed(1)
data("fake.f2",package="qtl")
fake_f2 <- qtl2::convert2cross2(fake.f2)
fake_f2 <- subset(fake_f2,chr=c(1:19))
smap_f2 <- cbind(runif(qtl2::n_ind(fake_f2),1,100),runif(qtl2::n_ind(fake_f2),1,100))
gmap_f2 <- qtl2::insert_pseudomarkers(fake_f2$gmap, step=1)
genoprobs_f2 <- qtl2::calc_genoprob(fake_f2,gmap_f2,error_prob=0.002)
eff_f2_1 <- eff_neighbor(genoprobs=genoprobs_f2,
                           pheno=fake_f2$pheno[,1],
                           gmap=gmap_f2, smap=smap_f2,
                           scale=20, addcovar=as.matrix(fake_f2$covar)
                           )
}
\references{
\itemize{
\item Haley CS, Knott SA (1992) A simple regression method for mapping quantitative trait loci in line crosses using flanking markers. Heredity 69:315-324.
\item Jansen RC (1993) Interval mapping of multiple quantitative trait loci. Genetics 135:205-211.
}
}
\author{
Yasuhiro Sato (\email{sato.yasuhiro.36c@kyoto-u.jp})
}
