% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/scan_neighbor.R
\name{scan_neighbor}
\alias{scan_neighbor}
\title{Genome scan for neighbor effects with a QTL model}
\usage{
scan_neighbor(
  genoprobs,
  pheno,
  gmap,
  contrasts = c(TRUE, TRUE, TRUE),
  smap,
  scale,
  addcovar = NULL,
  addQTL = NULL,
  grouping = rep(1, nrow(smap)),
  response = "quantitative"
)
}
\arguments{
\item{genoprobs}{Conditional genotype probabilities as obtained from \code{qtl2::calc_genoprob()}.}

\item{pheno}{A vector of individual phenotypes.}

\item{gmap}{Genetic map including observed and pseudomarkers, as obtained from \code{qtl2::insert_pseudomarkers()}.}

\item{contrasts}{A vector composed of three TRUE/FALSE values. Depending on crossing design, it represents the presence/absence of specific genotypes as c(TRUE/FALSE, TRUE/FALSE, TRUE/FALSE) = AA, AB, BB.}

\item{smap}{A matrix showing a spatial map. The first and second column include spatial points along a x-axis and y-axis, respectively.}

\item{scale}{A numeric scalar indicating the maximum spatial distance between a focal individual and neighbors to define neighbor effects.}

\item{addcovar}{An optional matrix including additional non-genetic covariates. It contains no. of individuals x no. of covariates.}

\item{addQTL}{An optional vector containing marker names that are considered covariates. Namely, this option allows composite interval mapping (Jansen 1993).}

\item{grouping}{An optional integer vector assigning each individual to a group. This argument can be useful when \code{smap} contains different experimental replicates. Default setting means that all individuals are belong to a single group.}

\item{response}{An optional argument to select trait types. The \code{"quantitative"} or \code{"binary"} calls the \code{"gaussian"} or \code{"binomial"} family in \code{glm()}, respectively.}
}
\value{
A matrix of LOD scores for self and neighbor effects, with the chromosome numbers and positions. The row names correspond to marker names.
\itemize{
\item{\code{chr}} {Chromosome number}
\item{\code{pos}} {Marker position}
\item{\code{LOD_self}} {LOD score for self effects}
\item{\code{LOD_nei}} {LOD score for neighbor effects}
}
}
\description{
Genome scan with a QTL model by Haley-Knott regression for self and neighbor effects, with possible allowance for additional covariates and non-normal traits.
}
\examples{
set.seed(1234)
data("fake.f2",package="qtl")
fake_f2 <- qtl2::convert2cross2(fake.f2)
fake_f2 <- subset(fake_f2,chr=c(1:19))
smap_f2 <- cbind(runif(qtl2::n_ind(fake_f2),1,100),runif(qtl2::n_ind(fake_f2),1,100))
gmap_f2 <- qtl2::insert_pseudomarkers(fake_f2$gmap, step=2)
genoprobs_f2 <- qtl2::calc_genoprob(fake_f2,gmap_f2)
s_seq <- quantile(dist(smap_f2),c(0.1*(1:10)))
scan_f2 <- scan_neighbor(genoprobs=genoprobs_f2,
                         pheno=fake_f2$pheno[,1], gmap=gmap_f2,
                         contrasts = c(TRUE,TRUE,TRUE), smap=smap_f2,
                         scale=19.37, addcovar=as.matrix(fake_f2$covar))
plot_nei(scan_f2)
}
\references{
\itemize{
\item Haley CS, Knott SA (1992) A simple regression method for mapping quantitative trait loci in line crosses using flanking markers. Heredity 69:315-324.
\item Jansen RC (1993) Interval mapping of multiple quantitative trait loci. Genetics 135:205-211.
}
}
\author{
Yasuhiro Sato (\email{sato.yasuhiro.36c@kyoto-u.jp})
}
