---
title: "rNeighborQTL"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{rNeighborQTL}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Overview
<p>The "rNeighborQTL" package includes core functions to perform a QTL mapping of neighbor effects. The "calc_neiprob()" computes conditional neighbor genotype identity using conditional genotype probabilities, and "scan_neighbor()" conducts interval mapping of neighbor effects. As an argument, the "calc_neiprob()" requires conditional genotype probabilities provided by the "r/qtl2" package (Broman et al. 2019). With genotypes, phenotypes, and spatial map (i.e., spatial position along x-axis and y-axis), the "scan_neighbor()" calculates LOD score from the conditional self-genotype probabilities and neighbor genotype identity.</p>

## Input files
<p>First of all, let us import genotypes and phenotypes using the "r/qtl" or "r/qtl2" package (Broman et al. 2003; Broman et al. 2019). Here is an example to import .csv files into a 'cross' object with "r/qtl" package, and covert it into a 'cross2' object with "r/qtl2" package. In this example, we import insect herbivory data on Col x Kas recombinant inbred lines of *Arabidopsis thaliana* (Wilson et al. 2001).</p>
```{r input_files}
colkas <- qtl::read.cross(format="csvs",dir=".",
                    genfile="../inst/ColKasGeno2001.csv",
                    phefile = "../inst/ColKas_fakePheno.csv",
                    na.strings = c("_"), estimate.map=TRUE)

colkas <- qtl2::convert2cross2(colkas)
```
  
<p>Subsequently, we obtain conditional self-genotype probabilities for pseudo-markers. For this purpose, we can use the "r/qtl2" package as follows.</p>
```{r get_selfprob}
gmap_colkas <- qtl2::insert_pseudomarkers(colkas$gmap, step=1)
colkas_genoprob <- qtl2::calc_genoprob(colkas,gmap_colkas,error_prob=0.002)
```

## Conditional neighbor genotypic identity
<p>Then, let us import the "rNeighborQTL" package and see how to test neighbor effects. Based on the self-genotype probabilities, here we calculate conditional neighbor genotypic identity</p>. 
```{r neiprob}
library(rNeighborQTL)
set.seed(1)
x <- runif(qtl2::n_ind(colkas),1,100)
y <- runif(qtl2::n_ind(colkas),1,100)
smap_colkas <- data.frame(x,y)

colkas_neiprob <- calc_neiprob(genoprobs=colkas_genoprob,
                              gmap=gmap_colkas, contrasts=c(TRUE,TRUE,TRUE),
                              smap=smap_colkas, scale=33,
                              a2=1, d2=0)
```
<p>where the arguments $a_2$ and $d_2$ indicates additive and dominance deviation of neighbor QTL effects, respectively.</p>

## Proportion of variation explained by neighbor effects
<p>Prior to the genome scan, we estimate the 'scale' argument. The "calc_pve()" computes proportion of phenotypic variation (PVE) by neighbor effects for a series of spatial scales. Based on the series of PVE, we calculate $\Delta$PVE and seek the scale $s$ that gives an argument for the maximum of $\Delta$PVE.</p>
```{r calc_pve}
s_seq <- quantile(dist(smap_colkas),c(0.1*(1:10))) #get quantiles for pairwise distances
colkas_pve <- calc_pve(genoprobs=colkas_genoprob, pheno=colkas$pheno[,2],
                       gmap=gmap_colkas, contrasts=c(TRUE,TRUE,TRUE),
                       smap=smap_colkas, s_seq=s_seq)
```

## QTL effects
<p>The deviation coefficients $a_2$ and $d_2$ are estimated using a linear regression on the genotype probabilities, also known as Haley-Knott regression (Haley & Knott 1992). The "eff_neighbor()" estimates the coefficients for self and neighbor effects, and plots the results as follows.</p>
```{r eff_neighbor, fig.width=4, fig.height=8}
colkas_effect1 <- eff_neighbor(genoprobs=colkas_genoprob,
                                  pheno=colkas$pheno[,2],
                                  gmap=gmap_colkas, contrasts=c(TRUE,TRUE,TRUE),
                                  smap=smap_colkas, scale=33, fig=TRUE)
```

## LOD score
<p>Lastly, we perform a genome scan to calculate LOD score for neighbor QTL effects. The "scan_neighbor()" calculates likelihoods using the estimated QTL effects through the "eff_neighbor()". The results can be shown by "plot_nei()".</p>
```{r scan_neighbor}
colkas_scan1 <- scan_neighbor(genoprobs=colkas_genoprob,
                              pheno=colkas$pheno[,2],
                              gmap=gmap_colkas, contrasts=c(TRUE,TRUE,TRUE),
                              smap=smap_colkas, scale=33)
plot_nei(colkas_scan1, type="neighbor")
```

## Options

### *1. Genome-wide threshold for the LOD score*
<p>In addition to the genome scan, we can perform permutation tests to add a genome-wide LOD threshold. Such permutation tests better account data structure, but require much computational cost.</p> 
```{r perm_neighbor}
perm_colkas <- perm_neighbor(genoprobs=colkas_genoprob, pheno=colkas$pheno[,2],
                              gmap=gmap_colkas, contrasts=c(TRUE,TRUE,TRUE),
                              smap=smap_colkas, scale=33,
                              times=9, p_val=0.01)
plot_nei(colkas_scan1, type="neighbor", th=perm_colkas$LOD_th)
```

<p>If the permutation tests are unrealistic, we instead apply a fast but conservative threshold using the likelihood ratio test with Bonferroni correction. The LOD threshold is determined by $\log_{10}(\exp(\chi^2))$, where the $\chi^2$ statistic and its degree-of-freedom is given by $p$-value divided by the number of markers and the number of model parameters, respectively.</p>
```{r bonf_neighbor}
plot_nei(colkas_scan1, type="neighbor", th="bonf",ylim=c(0,4.5))
```


### *2. Self-genotype effects*
<p>The "scan_neighbor()" also provides LOD score for self effects. This result is same as standard QTL mapping.</p>
```{r self}
plot_nei(colkas_scan1, type="self")
colkas_scan1 <- qtl2::scan1(colkas_genoprob,pheno=colkas$pheno[,2])
plot(colkas_scan1, map=gmap_colkas)
```

### *3. Composite interval mapping*
<p>The "addQTL" argument allows us to include non-focal QTLs as covariates. This option enables composite interval mapping (Jensen et al. 1993) that eliminates the additional QTL effects. Note that standard single-QTL analyses are applied for the covariate markers. Here is an example using the Col x Kas herbivory data.</p>
```{r CIM}
colkas_scan2 <- scan_neighbor(genoprobs=colkas_genoprob,
                              pheno=colkas$pheno[,2],
                              gmap=gmap_colkas, contrasts=c(TRUE,TRUE,TRUE),
                              smap=smap_colkas, scale=33,
                              addQTL=c("nga59","nga280"))
plot_nei(colkas_scan2)
```

### *4. Generalized linear models*
<p>The "response" argument allows us to analyze non-normal traits. The "quantitative", "binary", and "count" response links to Gaussian, binomial, and Poisson family in generalized linear models, respectively (Faraway 2016). Here is an example to analyze the number of leaf holes as a count variable using a Poisson error structure.</p>
```{r glm}
colkas_scan3 <- scan_neighbor(genoprobs=colkas_genoprob,
                              pheno=colkas$pheno[,3],
                              gmap=gmap_colkas, contrasts=c(TRUE,TRUE,TRUE),
                              smap=smap_colkas, scale=33,
                              response="count")
plot_nei(colkas_scan3)
```
<p>The "binary" response is also supported by the "calc_pve()" as it can call a logistic mixed model (Chen et al. 2016).</p>

### *5. Crossing design*
<p>Of course, heterozygotes are not expected in inbred lines. We can discard heterozygotes AB as NA, and perform a genome scan for inbred lines as follows.</p>
```{r inbred}
colkas <- qtl::read.cross(format="csvs",dir=".",
                      genfile="../inst/ColKasGeno2001_inbred.csv",
                      phefile = "../inst/ColKas_fakePheno.csv",
                      na.strings = c("_"), estimate.map=TRUE,
                      crosstype = "riself")

colkas <- qtl2::convert2cross2(colkas)
gmap_colkas <- qtl2::insert_pseudomarkers(colkas$gmap, step=1)
colkas_genoprob <- qtl2::calc_genoprob(colkas,gmap_colkas,error_prob=0.002)

colkas_scan4 <- scan_neighbor(genoprobs=colkas_genoprob,
                              pheno=colkas$pheno[,3],
                              gmap=gmap_colkas, contrasts=c(TRUE,FALSE,TRUE),
                              smap=smap_colkas, scale=33,
                              response="count")
plot_nei(colkas_scan4)
```
<p>We can also scan neighbor QTL effects even if there are only AA or AB genotypes. Note also that the "rNeighborQTL" does not support sex chromosomes currently, and they should be excluded before the genome scan. Examples using fake datasets (Broman et al. 2003) are shown in the "scan_neighbor()".</p>
```{r fake}
?scan_neighbor
```

## References
- Broman KW, Wu H, Sen S, Churchill GA. 2003. R/qtl: QTL mapping in experimental crosses. Bioinformatics 19: 889-890.
- Broman KW, Sen S, 2009. Single-QTL analysis, In: A guide to QTL mapping with R/qtl. Springer New York, New York, NY, pp. 75-133.
- Broman KW, Gatti DM, Simecek P, Furlotte NA, Prins P, Sen S, Yandell BS, Churchill GA. 2019. R/qtl2: Software for mapping quantitative trait loci with high-dimensional data and multiparent populations. Genetics 211: 495-502.
- Chen H, Wang C, Conomos M. et al. 2016. Control for population structure and relatedness for binary traits in genetic association studies via logistic mixed models. The American Journal of Human Genetics 98: 653-666.
- Faraway JJ. 2016. Extending the linear model with R: generalized linear, mixed effects and nonparametric regression models. CRC press.
- Haley CS, Knott SA. 1992. A simple regression method for mapping quantitative trait loci in line crosses using flanking markers. Heredity 69: 315-324.
- Jansen RC. 1993. Interval mapping of multiple quantitative trait loci. Genetics 135: 205-211.
- Wilson IW, Schiff CL, Hughes DE, Somerville SC. 2001. Quantitative trait loci analysis of powdery mildew disease resistance in the *Arabidopsis thaliana* accession kashmir-1. Genetics 158: 1301-1309.


